### ISSUE TITLE
When should delete respond with NOT_FOUND?

### ISSUE THREAD

**Issue by Siecje** [beanstalkd/beanstalkd#653]
In the description for delete when describing when the server will return `NOT_FOUND\r\n` it says

> This could happen if the job timed out before the
   client sent the delete command.

I assume this means the client has reserved the job and sent `delete` but the job ttr was reached.

But if the server released the job (and there are no other clients waiting for a job) then the job will be in the ready state and the job will be deleted and the response would be `DELETED\r\n`.

Am I missing something?

https://github.com/beanstalkd/beanstalkd/blob/91c54fc05dc759ef27459ce4383934e1a4f2fb4b/doc/protocol.txt#L271-L286

**Comment by georaldc** [comment]
Maybe the wording needs to be updated? Looking at the source, it seems like several attempts are made to delete a job id, going through the following possible states: reserved, ready, buried, delayed. If none of those go through, then it returns the NOT FOUND message. Looking at that, it seems the only reason you would get NOT FOUND is if the job no longer exists at that point (removed by another process?).

https://github.com/beanstalkd/beanstalkd/blob/91c54fc05dc759ef27459ce4383934e1a4f2fb4b/prot.c#L1516

**Comment by Siecje** [comment]
What if the job is current reserved by another client?

**Comment by georaldc** [comment]
> What if the job is current reserved by another client?

I tested this and the prior client will get a not found error. Basically, only one client (the last client to reserve a specific job) may modify a reserved job's state, or delete a job. I assume this can happen if TTR was reached in the middle of the prior client's processing, or if beanstalkd was restarted with binlogs enabled (the reserved job will return back into a ready state) and another client reserves the job.

If a delete request against the job ID was sent before another client gets the chance to reserve it, then yes the delete will go through (at least from what I found during a quick test)
